File: LEETCODE/Problem1/TwoSum.py
# Runtime: 48 ms, faster than 99.13% of Python3 online submissions for Two Sum.
# Memory Usage: 15.4 MB, less than 20.75% of Python3 online submissions for Two Sum.

class Solution:
    def twoSum(self, nums, target):
        visited={}
        for i in range(len(nums)):
            remain=target-nums[i]
            if remain in visited.keys():
                return [visited[remain],i]
            visited[nums[i]]=i
            


sol=Solution()
print(sol.twoSum([3,2,4],6))




# runs with 30% better tc and  80 % better sc


# class Solution:
#     def twoSum(self, nums, target):
#         ans=[]
#         count=0
#         for i,x in enumerate(nums):
#             if target - x in nums[:i]+nums[i+1:]:
#                 target =target-x
#                 ans.append(i)
#                 p=i+1
#                 ans.append(nums[p:].index(target)+p)
#                 break
#         return ans
                
            


# sol=Solution()
# print(sol.twoSum([2,5,5,11],10))








# [-3,4,3,90]
# 0


# [0,1]

File: LEETCODE/Problem1/TwoSum.cpp
// Runtime: 8 ms, faster than 94.73% of C++ online submissions for Two Sum.
// Memory Usage: 11.2 MB, less than 22.27% of C++ online submissions for Two Sum.

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        
        map<int,int> mp;
        int rem;

        for(int i=0;i<nums.size();i++)
        {
            rem=target-nums.at(i);
            if(mp.count(rem))
            {
                return {mp.at(rem),i};
            }
            mp[nums.at(i)]=i;
        }

        return {};
    }
};


int main()
{
    vector<int> vect={3,3};
    int target=6;
    Solution solution=Solution();
    vector<int> ans=solution.twoSum(vect,target);
    for(int i=0;i<ans.size();i++)
    {
        cout<<ans.at(i)<<endl;
    }
}